tee/prototype/epochproc.go

```diff
@@ -0,0 +1,397 @@
+func (e *Enclave) txRoutine(
+	phaseShift chan<- struct{},
+	exits <-chan exitersSet,
+	txs <-chan *tee.Transaction,
+	txEpoch *Epoch,
+) error {
+	var stagedTxs txCache
+	for {
+		bpCache, err := e.generateBalanceProofs(txEpoch)
+		if err != nil {
+			return fmt.Errorf("generating balance proofs: %w", err)
+		}
+		select {
+		case exiters := <-exits:
+			txEpoch, bpCache, ptxs := prune(txEpoch, stagedTxs, bpCache, exiters)
+			bpCache, err = e.applyEpochTx(txEpoch, ptxs, bpCache)
+			if err != nil {
+				return fmt.Errorf("adjusting Epoch %v Balances: %w", txEpoch.Number, err)
+			}
+			e.balanceProofs <- getBalProofs(bpCache)
+			phaseShift <- struct{}{} // phase done, stop processing TXs and signal epoch shift.
+			return nil
+		case tx := <-txs:
+			stagedTxs = cacheTx(stagedTxs, tx)
+		}
+	}
+}
+
+// getBalProofs retrieves a slice of `*tee.BalanceProof`s from the balance proof cache.
+func getBalProofs(bps balProofCache) []*tee.BalanceProof {
+	proofs := make([]*tee.BalanceProof, len(bps))
+	i := 0
+	for _, bp := range bps {
+		proofs[i] = bp
+	}
+	return proofs
+}
+
+// prune prunes a given slice of transactions for a given set of addresses, s.t.
+// no element of the given set is either a recipient or sender in a transaction,
+// it also makes sure to remove exited parties from the txEpoch and the
+// balance proof cache.
+func prune(epoch *Epoch,
+	txs txCache,
+	bpc balProofCache,
+	forS []common.Address,
+) (*Epoch, balProofCache, []*tee.Transaction) {
+	for _, m := range forS {
+		delete(epoch.balances, m) // remove member from txEpoch
+		delete(bpc, m)            // remove member from balProofCache
+
+		sTxs, sOk := txs.senders[m]
+		rTxs, rOk := txs.recipients[m]
+		if !sOk && !rOk {
+			continue
+		}
+		for _, tx := range sTxs {
+			delete(txs.txs, tx.Hash())
+		}
+		for _, tx := range rTxs {
+			delete(txs.txs, tx.Hash())
+		}
+	}
+	prunedTxs := make([]*tee.Transaction, 0, len(txs.txs))
+	i := 0
+	for _, tx := range txs.txs {
+		prunedTxs[i] = tx
+	}
+	return epoch, bpc, prunedTxs
+}
+
+type balProofCache = map[common.Address]*tee.BalanceProof
+
+func (e *Enclave) depositExitRoutine(
+	verifiedBlocks <-chan *tee.Block,
+	phaseDone chan<- exitersSet,
+	depositEpoch, exitEpoch *Epoch,
+) error {
+	// read blocks from verifiedBlocks (deposit phase).
+	for vb := range verifiedBlocks {
+		exiters, err := e.handleVerifiedBlock(depositEpoch, exitEpoch, vb)
+		if err != nil {
+			return fmt.Errorf("handling verified blocknr %v : %w", vb.NumberU64(), err)
+		}
+		if e.phaseDone(vb.NumberU64()) {
+			e.depositProofs <- retrieveCachedDepProofs(e.depositProofCache)
+			e.depositProofCache = make(map[common.Address]*tee.DepositProof)
+			phaseDone <- exiters
+			close(phaseDone) // phase done, stop TX processor.
+			return nil
+		}
+	}
+	return nil
+}
+
+// retrieveCachedDepProofs reduces the deposit proof cache to a slice of `tee.DepositProof`s.
+func retrieveCachedDepProofs(cache map[common.Address]*tee.DepositProof) []*tee.DepositProof {
```

Comment:
This func and `getBalProofs` do conceptually the same, maybe we call them similarly? like `asFooProofs`? Also, in one function you use `append` in the other you address the slice index.